content/_code-samples/key-derivation/key-derivation.py

```diff
@@ -0,0 +1,327 @@
+        This is the public key used for validators.
+        """
+        if self._secp256k1_root_pub is None:
+            self.derive_secp256k1_master_keys()
+        return self._secp256k1_root_pub
+
+    def derive_secp256k1_master_keys(self):
+        """
+        Uses the XRPL's convoluted key derivation process to get the
+        secp256k1 master keypair for this seed value.
+        Saves the values to the object for later reference.
+        """
+
+        root_sec_i = secp256k1_secret_key_from(self.bytes)
+        root_pub_point = keys.get_public_key(root_sec_i, curve.secp256k1)
+        root_pub_b = compress_secp256k1_public(root_pub_point)
+        fam_b = bytes(4) # Account families are unused; just 4 bytes of zeroes
+        inter_pk_i = secp256k1_secret_key_from(root_pub_b+fam_b)
+        inter_pub_point = keys.get_public_key(inter_pk_i, curve.secp256k1)
+
+        # Secret keys are ints, so just add them mod the secp256k1 group order
+        master_sec_i = (root_sec_i + inter_pk_i) % SECP_MODULUS
+        # Public keys are points, so the fastecdsa lib handles adding them
+        master_pub_point = root_pub_point + inter_pub_point
+
+        self._secp256k1_sec = master_sec_i.to_bytes(32, byteorder="big", signed=False)
+        self._secp256k1_pub = compress_secp256k1_public(master_pub_point)
+        self._secp256k1_root_pub = root_pub_b
+
+        # Saving the full key to make it easier to sign things later
+        self._secp256k1_full = master_pub_point
+
+    def encode_secp256k1_public_base58(self, validator=False):
+        """
+        Return the base58-encoded version of the secp256k1 public key.
+        """
+        if validator:
+            # Validators use the "root" public key
+            key = self.secp256k1_root_public_key
+            prefix = XRPL_VALIDATOR_PUBKEY_PREFIX
+        else:
+            # Accounts use the derived "master" public key
+            key = self.secp256k1_public_key
+            prefix = XRPL_ACCT_PUBKEY_PREFIX
+
+        return base58.b58encode_check(prefix + key).decode()
+
+    def encode_ed25519_public_base58(self):
+        """
+        Return the base58-encoded version of the Ed25519 public key.
+        """
+        # Unlike secp256k1, Ed25519 public keys are the same for
+        # accounts and for validators.
+        prefix = XRPL_ACCT_PUBKEY_PREFIX
+
+        return base58.b58encode_check(prefix +
+                                      self.ed25519_public_key).decode()
+
+def secp256k1_secret_key_from(seed):
+    """
+    Calculate a valid secp256k1 secret key by hashing a seed value;
+    if the result isn't a valid key, increment a seq value and try
+    again.
+
+    Returns a secret key as a 32-byte integer.
+    """
+    seq = 0
+    while True:
+        buf = seed + seq.to_bytes(4, byteorder="big", signed=False)
+        h = sha512half(buf)
+        h_i = int.from_bytes(h, byteorder="big", signed=False)
+        if h_i > SECP_MODULUS or h_i == 0:
```

Comment:
We should check for greater than or equal to, not just greater than. I.e `h_i >= SECP_MODULUS`